test/integration/apimachinery/admission_test.go

```diff
@@ -0,0 +1,871 @@
+			return false
+		}
+		if gvrsToTest[i].Resource < gvrsToTest[j].Resource {
+			return true
+		}
+		if gvrsToTest[i].Resource > gvrsToTest[j].Resource {
+			return false
+		}
+		return true
+	})
+
+	// Test admission on all resources, subresources, and verbs
+	for _, gvr := range gvrsToTest {
+		resource := resourcesByGVR[gvr]
+		t.Run(gvr.Group+"."+gvr.Version+"."+resource.Name, func(t *testing.T) {
+			for _, verb := range []string{"create", "update", "patch", "connect", "delete", "deletecollection"} {
+				if shouldTestResourceVerb(gvr, resource, verb) {
+					t.Run(verb, func(t *testing.T) {
+						holder.reset(t)
+						testFunc := getTestFunc(gvr, verb)
+						testFunc(&testContext{
+							t:               t,
+							admissionHolder: holder,
+							client:          dynamicClient,
+							gvr:             gvr,
+							resource:        resource,
+							resources:       resourcesByGVR,
+						})
+						holder.verify(t)
+					})
+				}
+			}
+		})
+	}
+}
+
+//
+// generic resource testing
+//
+
+func testResourceCreate(c *testContext) {
+	stubObj, err := getStubObj(c.gvr, c.resource)
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+	ns := ""
+	if c.resource.Namespaced {
+		ns = testNamespace
+	}
+	c.admissionHolder.expect(c.gvr, schema.GroupVersionKind{c.resource.Group, c.resource.Version, c.resource.Kind}, v1beta1.Create, stubObj.GetName(), ns, true, false)
+	_, err = c.client.Resource(c.gvr).Namespace(ns).Create(stubObj, metav1.CreateOptions{})
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+}
+
+func testResourceUpdate(c *testContext) {
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		obj, err := createOrGetResource(c.client, c.gvr, c.resource)
+		if err != nil {
+			return err
+		}
+		obj.SetAnnotations(map[string]string{"update": "true"})
+		c.admissionHolder.expect(c.gvr, schema.GroupVersionKind{c.resource.Group, c.resource.Version, c.resource.Kind}, v1beta1.Update, obj.GetName(), obj.GetNamespace(), true, true)
+		_, err = c.client.Resource(c.gvr).Namespace(obj.GetNamespace()).Update(obj, metav1.UpdateOptions{})
+		return err
+	}); err != nil {
+		c.t.Error(err)
+		return
+	}
+}
+
+func testResourcePatch(c *testContext) {
+	obj, err := createOrGetResource(c.client, c.gvr, c.resource)
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+	c.admissionHolder.expect(c.gvr, schema.GroupVersionKind{c.resource.Group, c.resource.Version, c.resource.Kind}, v1beta1.Update, obj.GetName(), obj.GetNamespace(), true, true)
+	_, err = c.client.Resource(c.gvr).Namespace(obj.GetNamespace()).Patch(
+		obj.GetName(),
+		types.MergePatchType,
+		[]byte(`{"metadata":{"annotations":{"patch":"true"}}}`),
+		metav1.PatchOptions{})
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+}
+
+func testResourceDelete(c *testContext) {
+	obj, err := createOrGetResource(c.client, c.gvr, c.resource)
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+	background := metav1.DeletePropagationBackground
```

Comment:
We should also test the case where finalizer is added. That exercises a different code path leading to webhook calls (https://github.com/kubernetes/kubernetes/pull/76346/files#diff-2c081fad5c858e67c96f75adac185093R786).

Or you can leave this to me in #76346.

FYI, the "delete on update" case doesn't invoke the webhook differently, so no need to add that test to testResourceUpdate.

Reply:
added

Reply:
I'll let you tweak the delete test with finalizer in your PR